### PR TITLE
Support ON CONFLICT DO NOTHING on compressed hypertables

### PR DESCRIPTION
**Pull request by svenklemm**
This patch allows unique constraints on compressed chunks. When trying to INSERT into compressed chunks with unique constraints any potentially conflicting compressed batches will be decompressed to let postgres do constraint checking on the INSERT. With this patch only INSERT ON CONFLICT DO NOTHING will be supported. For decompression only segment by information is considered to determine conflicting batches. This will be enhanced in a follow-up patch to also include orderby metadata to require decompressing less batches.

### CODE REVIEW

File: src/nodes/chunk_dispatch/chunk_dispatch.c [hunk 1]
```diff
@@ -106,8 +108,27 @@ ts_chunk_dispatch_get_chunk_insert_state(ChunkDispatch *dispatch, Point *point,
 			chunk = ts_hypertable_create_chunk_for_point(dispatch->hypertable, point, &found);
 		}
 		else
+		{
 			found = true;
 
+			if (ts_chunk_is_compressed(chunk) && chunk->data_nodes == NIL && slot)
```
Comment (akuzm) [inline src/nodes/chunk_dispatch/chunk_dispatch.c:114]: slot can't be null, right? Might be more clear to convert the above Assert into Ensure if it's just to prevent segfaults

File: tsl/src/compression/compression.c [hunk 2]
```diff
@@ -814,6 +817,7 @@ index_scan_sequence_number(Relation table_rel, Oid index_oid, ScanKeyData *scank
 	index_scan->xs_want_itup = true;
 	index_rescan(index_scan, scankey, num_scankeys, NULL, 0);
 
+	/* TODO: add sanity check */
```
Comment (akuzm) [inline tsl/src/compression/compression.c:820]: Can you expand on this?

File: tsl/src/compression/compression.c [hunk 3]
```diff
@@ -2042,3 +2048,199 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
...
+
+	RowDecompressor decompressor = build_decompressor(in_rel, out_rel);
+	Bitmapset *unique_columns = relation_unique_columns(out_rel);
+
+	int num_scankeys;
+	ScanKeyData *scankeys =
+		build_scankeys(chunk->fd.hypertable_id, decompressor, unique_columns, slot, &num_scankeys);
+
+	TableScanDesc heapScan =
+		table_beginscan(in_rel, GetTransactionSnapshot(), num_scankeys, scankeys);
```
Comment (akuzm) [inline tsl/src/compression/compression.c:2204]: This is a seq scan, probably we want to have an index scan?

File: tsl/src/compression/compression.c [hunk 4]
```diff
@@ -2042,3 +2048,199 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
...
+
+	for (HeapTuple compressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
```
Comment (akuzm) [inline tsl/src/compression/compression.c:2206]: Suspiciously low coverage for this loop, is it codecov mistake or missing tests?

File: tsl/src/compression/compression.c [hunk 5]
```diff
@@ -2042,3 +2048,199 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
...
+				TypeCacheEntry *tce =
+					lookup_type_cache(decompressor.in_desc->attrs[attno - 1].atttypid,
+									  TYPECACHE_EQ_OPR_FINFO);
+
+				if (!tce || !OidIsValid(tce->eq_opr_finfo.fn_oid))
+					elog(ERROR,
+						 "no equality function for type \"%s\"",
+						 format_type_be(decompressor.in_desc->attrs[attno - 1].atttypid));
+
+				if (isnull)
```
Comment (akuzm) [inline tsl/src/compression/compression.c:2147]: Would be nice to have coverage for null value in segmentby column as well.

File: tsl/src/compression/compression.c (hunk 3 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2204]: We need to support both, you cant have scankeys on non-index column in indexscan, right?
But indexscan support will not be part of initial PR

Comment (akuzm) [reply tsl/src/compression/compression.c:2204]: Right. But Postgres requires there to be an index for the ON CONFLICT columns, so we should have one. I think it's OK with seq scan for starters, but performance won't be good and we probably need to add index scan soon.

Comment (akuzm) [reply tsl/src/compression/compression.c:2204]: Oh, it's more difficult with orderby columns, we don't have min/max in the default index for them.

File: src/nodes/chunk_dispatch/chunk_dispatch.c (hunk 1 above)
Comment (svenklemm) [reply src/nodes/chunk_dispatch/chunk_dispatch.c:114]: there is an assert earlier that is never hit and i think that is leftover from an earlier version

File: tsl/src/compression/compression.c (hunk 2 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c]: that isnt even code added by this PR, removed

File: tsl/src/compression/compression.c (hunk 3 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2203]: the index will be on the uncompressed part though, not the compressed part

File: tsl/src/compression/compression.c (hunk 4 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2205]: tests are WIP

File: tsl/src/compression/compression.c (hunk 5 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2155]: done

File: tsl/src/compression/compression.c [hunk 6]
```diff
@@ -2042,3 +2047,226 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
...
+
+		heap_deform_tuple(compressed_tuple,
+						  decompressor.in_desc,
+						  decompressor.compressed_datums,
+						  decompressor.compressed_is_nulls);
+
+		row_decompressor_decompress_row(&decompressor);
+
+		TM_FailureData tmfd;
+		table_tuple_delete(in_rel,
```
Comment (erimatnor) [inline tsl/src/compression/compression.c:2246]: Suggest handling return value here, or at least asserting the result if only one possibility.

File: tsl/src/compression/compression.c [hunk 7]
```diff
@@ -2042,3 +2047,226 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
...
+										   decompressor,
+										   unique_columns,
+										   &null_columns,
+										   slot,
+										   &num_scankeys);
+
+	TableScanDesc heapScan =
+		table_beginscan(in_rel, GetTransactionSnapshot(), num_scankeys, scankeys);
+
+	for (HeapTuple compressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
```
Comment (erimatnor) [inline tsl/src/compression/compression.c:2211]: Heap scans like this without a per-tuple memory context make me suspicious. I worry that there's a risk this allocates memory proportional to the size of the table. You should double check the memory usage, especially for big tables.

File: src/copy.c [hunk 8]
```diff
@@ -324,6 +324,7 @@ TSCopyMultiInsertBufferFlush(TSCopyMultiInsertInfo *miinfo, TSCopyMultiInsertBuf
 	ChunkInsertState *cis =
 		ts_chunk_dispatch_get_chunk_insert_state(miinfo->ccstate->dispatch,
 												 buffer->point,
+												 buffer->slots[0],
```
Comment (akuzm) [inline src/copy.c:327]: This looks incorrect, we should have decompressed everything we needed. Let's move decompression out of this function? It looks error-prone in cases like this one, when you just want to re-create the insert state, but don't actually want to decompress anything. Or when you use some batching, get the chunk insert state only once, but then have to call decompression for every row.

File: tsl/src/compression/compression.c (hunk 7 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2211]: the row decompressor switches to a per tuple memory context

File: src/copy.c (hunk 8 above)
Comment (svenklemm) [reply src/copy.c:327]: Hmm in the presence of unique constraints you have to do this check for every tuple. Cause different tuples routed to the same chunk might still require different batches to be decompressed. Where would you move this to?

File: tsl/src/compression/compression.c (hunk 3 above)
Comment (akuzm) [reply tsl/src/compression/compression.c:2209]: Let's discuss the indexes a little more. I think the implementation with seq scan is OK to merge for starters, but it's not something we should publish. Doing a seq scan for each inserted row is going to be prohibitively slow.

I think it is reasonable for us to require an index on the decompressed chunk that matches the ON CONFLICT clause. It's the same thing that Postgres requires.

1) We already create the index on segmentby columns, so it always can be used.
2) For min/max, can probably use a seq scan?
3) Ideally, if Postgres managed to find an arbiter index on the uncompressed chunk, we should have a matching index on the decompressed chunk. Not sure if we do anything to guarantee that currently.

Do we have a plan for this somewhere?

File: src/copy.c (hunk 8 above)
Comment (akuzm) [reply src/copy.c:327]: To `chunk_dispatch_exec` and `copyfrom`, just after they look up the insert state. And here, in `TSCopyMultiInsertBufferFlush`, you flush the multi-insert buffers with tuples that you have already seen in `copyfrom`, so you don't have to do decompression for them again.

File: tsl/src/compression/compression.c (hunk 3 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2209]: There will never be a matching index if you have unique constraints. Since unique constraints need to include the partitioning columns that means time will be part of the index but you will never want time as segmentby .

File: tsl/src/compression/compression.c [hunk 9]
```diff
@@ -2042,3 +2047,226 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
...
+		ReleaseSysCache(index_tuple);
+	}
+
+	list_free(indexoidlist);
+
+	return unique_columns;
+}
+
+static ScanKeyData *
+build_scankeys(int32 hypertable_id, RowDecompressor decompressor, Bitmapset *unique_columns,
```
Comment (akuzm) [inline tsl/src/compression/compression.c:2099]: This part looks like "planning-time", similar to the deduction of arbiter indexes. But probably there's no convenient way to do it there. I think it would be good not do it per-row and cache this in the chunk insert state.

Comment (svenklemm) [reply tsl/src/compression/compression.c:2099]: This needs to be per row as the scankeys are built from row values. I dont think you can easily cache this.

Comment (akuzm) [reply tsl/src/compression/compression.c:2099]: Postgres index scans do that by just replacing the `sk_argument` (`ExecIndexEvalRuntimeKeys`).

**Review by akuzm — APPROVED** [review]

File: src/process_utility.c [hunk 10]
```diff
@@ -2667,7 +2667,7 @@ process_index_start(ProcessUtilityArgs *args)
 	hcache = ts_hypertable_cache_pin();
 	ht = ts_hypertable_cache_get_entry_rv(hcache, stmt->relation);
 
-	if (NULL == ht)
+	if (!ht)
```
Comment (konskov) [inline src/process_utility.c:2670]: nit: I think these changes as well as https://github.com/timescale/timescaledb/pull/5252/files#diff-038649a1ce33b21f89e080d566755f896778f29c2ae7c5d9a1bf6e7975b5f48cR1409 could be made separately in a different PR, if they are necessary?

File: tsl/src/compression/compression.c [hunk 11]
```diff
@@ -2042,3 +2047,226 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
...
+			 * we have no metadata.
+			 */
+
+			if (COMPRESSIONCOL_IS_SEGMENT_BY(fd))
+			{
+				bool isnull;
+				Datum value = slot_getattr(slot, attno, &isnull);
+
+				TypeCacheEntry *tce =
+					lookup_type_cache(decompressor.in_desc->attrs[attno - 1].atttypid,
```
Comment (konskov) [inline tsl/src/compression/compression.c:2138]: `attno` is taken from the uncompressed chunk. is it guaranteed to be the same for the compressed? or should we be using the attname again to get the attno for the compressed chunk?

File: tsl/src/compression/compression.c [hunk 12]
```diff
@@ -1503,6 +1507,7 @@ typedef struct RowDecompressor
 
 	TupleDesc out_desc;
 	Relation out_rel;
+	CatalogIndexState indexstate;
```
Comment (konskov) [inline tsl/src/compression/compression.c:1510]: this is not used anywhere yet, right?

File: tsl/src/compression/compression.c (hunk 11 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2138]: good find, we do indeed need to do proper mapping here

Comment (antekresic) [reply tsl/src/compression/compression.c:2138]: Any reason not to convert `attno - 1` to `AttrNumberGetAttrOffset(attno)`?

File: tsl/src/compression/compression.c (hunk 12 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:1510]: correct, removing

File: tsl/src/compression/compression.c (hunk 9 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2098]: i'll leave that for a followup PR

File: tsl/src/compression/compression.c (hunk 11 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c]: done

File: tsl/src/compression/compression.c (hunk 6 above)
Comment (svenklemm) [reply tsl/src/compression/compression.c:2191]: done

**Comment by konskov** [comment]
approving, I think this should be 1 commit though, right?

**Review by konskov — APPROVED** [review]
